app/Http/Controllers/AuthsignalController.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Models\User;
+use Authsignal;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+
+class AuthsignalController extends Controller
+{
+    public function settingsPage()
+    {
+        $result = Authsignal::getUser(
+            Auth::id()
+        );
+
+        return view('auth.settings', [
+            'enrolledVerificationMethods' => $result['enrolledVerificationMethods'] ?? []
+        ]);
+    }
+
+    public function enrollMFA(Request $request) {
+        $user = Auth::user();
+        $result = Authsignal::getUser(Auth::id());
+
+        $isEnrolled = $result['isEnrolled'];
+
+        $result = Authsignal::track($user->id, 'enroll', [
+            'redirectToSettings' => $isEnrolled,
+            'redirectUrl' => route('auth.validate-mfa-challenge', ['remember' => $request->boolean('remember')]),
+            'email' => $user->email,
+            'deviceId' => $request->cookie('__as_aid'),
+            'userAgent' => $_SERVER["HTTP_USER_AGENT"],
+            'ipAddress' => $request->ip(),
+        ]);
+
+        if ($result['state'] === 'ALLOW') {
+            return redirect($result['url']);
+        } else {
+            return redirect()->route('auth.settings')
+                ->with('status', 'Unable to enroll MFA. Please try again later.');
+        }
+    }
+
+    public function validateMfaChallenge(Request $request)
+    {
+        $token = $request->input('token');
+        if (!$token) {
+            abort(400, 'Token is required');
+        }
+
+        $user = Auth::user();
+        $response = Authsignal::validateChallenge($token, $user->id);
+
+        return redirect()->route('auth.settings')
+            ->with('status', $response['isValid'] ? 'Enrolled MFA successfully' : 'Failed to enroll MFA');
+    }
+
+    public function validateAuthenticationChallenge(Request $request)
+    {
+        if (!$request->hasValidSignatureWhileIgnoring(['token'])) {
+            abort(401);
+        }
+
+        $token = $request->get('token');
+        $userId = $request->get('userId');
+
+        if (!$token || !$userId) {
+            abort(400, 'token and userId are required');
+        }
+
+        $response = Authsignal::validateChallenge($token, $userId);
+
+        if ($response['isValid'] && $response["state"] === 'CHALLENGE_SUCCEEDED') {
```

Comment:
You can remove `&& $response["state"] === 'CHALLENGE_SUCCEEDED'` here as isValid won't be true if the state is not CHALLENGE_SUCCEEDED.